src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -227,32 +297,91 @@ private void OnDiagnosticsUpdatedOnBackground(DiagnosticsUpdatedArgs e)
             /// Similarly, clearing out data is just a matter of us clearing our reference
             /// to the data.  
             /// </summary>
-            private void OnDiagnosticsUpdatedOnForeground(DiagnosticsUpdatedArgs e)
+            private void OnDiagnosticsUpdatedOnForeground()
             {
-                this.AssertIsForeground();
+                ProviderAndDocumentToBatchedUpdates batchedUpdates;
+                lock (_gate)
+                {
+                    Debug.Assert(_updateTask != null);
 
-                if (_disposed)
+                    // Grab the work we need to do.  Create a fresh dictionary for new work to be
+                    // put into.
+                    batchedUpdates = _idToBatchedUpdates;
+                    _idToBatchedUpdates = s_providerPool.Allocate();
+
+                    // Clear out the task so that any new work that comes in will cause a new update
+                    // task to be created.
+                    _updateTask = null;
+                }
+
+                try
                 {
-                    return;
+                    if (_disposed)
+                    {
+                        return;
+                    }
+
+                    // Do some quick checks to avoid doing any further work for diagnostics  we don't
+                    // care about.
+                    var ourDocument = _subjectBuffer.AsTextContainer().GetOpenDocumentInCurrentContext();
+                    var ourDocumentId = ourDocument?.Id;
+                    if (ourDocumentId != _currentDocumentId)
```

Comment:
I would prefer doing this in BG by listening to all proper workspace events on buffer/workspace changes notifications. if those events are broken, we should fix those events.

Reply:
We could do that as a later change.  It will make things more complex as we'll have to serialize out our BG requests and filter them against the requests to update _currentDocId.  i'm ok doing that as part of a followup PR.

Reply:
Note: can you tell me precisely which Workspace/buffer changes I would need to listen to?

Reply:
ContextChange and WorkspaceChanged event.